tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner_test.py

```diff
@@ -0,0 +1,343 @@
+        dtype=weight_dtype)
+    sparse_vars = [(mask, weight, self.grad(weight))]
+    drop_ratio = 0.
+
+    p = pruner.RiGLPruner(
+      update_schedule=self.updater(drop_ratio, begin_step, end_step, freq_update),
+      sparsity=self.target_sparsity,
+      block_size=self.block_size,
+      block_pooling_type=self.block_pooling_type,
+      stateless=self.stateless,
+      seed=self.seed,
+      noise_std=self.noise_std,
+      reinit=self.reinit
+    )
+
+    optimizer = tf.keras.optimizers.SGD(learning_rate=0.01)
+    optimizer.iterations.assign(0)
+    p.create_slots(optimizer, weight)
+
+    for _ in tf.range(end_step + 2):
+      mask_before_update = optimizer.get_slot(weight, 'mask').read_value()
+      before_sparsity = np.count_nonzero(mask_before_update) / tf.size(mask_before_update)
+      step = optimizer.iterations
+      p.update_masks(sparse_vars, step)
+      mask_after_update = optimizer.get_slot(weight, 'mask').read_value()
+      after_sparsity = np.count_nonzero(mask_after_update) / tf.size(mask_before_update)
+      self.assertAllEqual(mask_before_update, mask_after_update)
+
+  @parameterized.parameters(
+    (0.7,), (0.1,), (0.3,), (0.5,), 
+  )
+  def testUpdatesAccordingtoSchedule(self, drop_ratio):
+    # unlike previous tests, this checks that the masked weights
+    # maintain consistent sparsity and that there is some change in the mask.
+    weight = tf.Variable(np.linspace(1.0, 100.0, 100))
+    weight_dtype = weight.dtype.base_dtype
+    mask = tf.Variable(
+        tf.ones(weight.get_shape(), dtype=weight_dtype),
+        dtype=weight_dtype)
+    grad = self.grad(weight)
+    sparse_vars = [(mask, weight, grad)]
+    sparsity_params = {
+      'pruning_schedule': self.updater(drop_ratio, 1, 4, 2)
+    } # dummy just for using the pruning optimizer
+    sparsity_config = pruning_config.LowMagnitudePruningConfig(**sparsity_params) # dummy
+
+    p = pruner.RiGLPruner(
+      update_schedule=self.updater(drop_ratio, 1, 4, 2), # update iter 1 and 3
+      sparsity=self.target_sparsity,
+      block_size=self.block_size,
+      block_pooling_type=self.block_pooling_type,
+      stateless=self.stateless,
+      seed=self.seed,
+      noise_std=self.noise_std,
+      reinit=self.reinit
+    )
+
+    _optimizer = tf.keras.optimizers.SGD(learning_rate=0.01)
+    optimizer = pruning_optimizer.PruningOptimizer(_optimizer, sparsity_config)
+    optimizer.iterations.assign(0)
+
+    p.create_slots(optimizer, weight)
+
+    def _train(optimizer, weight):
+      expected_iter1 = tf.zeros_like(weight, dtype=weight.dtype.base_dtype) * -1
+      expected_iter3 = tf.zeros_like(weight, dtype=weight.dtype.base_dtype) * -1
+      for i in tf.range(0, 5):
+        mask_before_update = optimizer.get_slot(weight, 'mask').read_value()
+        mask_before_update3 = optimizer.get_slot(weight, 'mask').read_value()
+        before_sparsity = tf.math.count_nonzero(mask_before_update, dtype=tf.int32) / tf.size(mask_before_update)
```

Comment:
Are you going to run into trouble w/ the int division happening here? Probably much safer to explicitly use floats

Reply:
yep, the default is int64 so it can't be downcasted.